align_system/utils/incontext_utils.py

```diff
@@ -0,0 +1,287 @@
+import json
+import torch
+import random
+from copy import deepcopy
+from abc import ABCMeta, abstractmethod
+
+from align_system.utils import adm_utils
+from align_system.utils import outlines_prompts_utils
+from align_system.utils.hydrate_state import hydrate_scenario_state
+from align_system.prompt_engineering.outlines_prompts import (
+    scenario_state_description_with_relevant_char_info,
+    comparative_kdma_score_prediction_prompt
+)
+
+
+class IncontextExampleGenerator(object, metaclass=ABCMeta):
+    '''
+    Abstract class for incontext example generator
+    Instances of this class have unique set_icl_datasets() functions for formatting prompt and reponses 
+    '''
+    @abstractmethod
+    def __init__(self,
+                 incontext_settings, 
+                 target_kdmas, 
+                 log,
+                 **kwargs):
+        self.incontext_settings = incontext_settings
+        self.target_kdmas = target_kdmas
+        self.log = log
+        self.set_icl_datasets()
+
+    @abstractmethod
+    def set_icl_datasets(self):
+        '''
+        Sets self.icl_datasets which contains all the ICL examples
+        This is specific to each instance of the class because the prompt and response will vary
+        read_icl_dataset_files() is a generic helper method for this step
+
+        The keys of self.icl_datasets are the 'kdma' keys from self.target_kdmas,
+        the values are a list of all ICL examples for that kdma,
+        each ICL example is a dictionary with keys: 'prompt', 'response'
+        For example: {kdma: [{'prompt':str, 'response':json}, ...], ...}
+        '''
+        self.icl_datasets = {}
+        pass
+
+    def read_icl_dataset_files(self):
+        '''
+        Helper function for set_icl_datasets() - reads dataset files and gets examples for target_kdmas
+        Returns incontext_data dictionary with format:
+            {kdma:[{state, actions, choices, kdma_values}, ...], ...}
+        '''
+        incontext_data = {}
+        # For each kdma
+        for target_kdma in self.target_kdmas:
+            sys_kdma_name = target_kdma['kdma']
+            # Add examples for each dataset file
+            dset_files = self.incontext_settings["datasets"][sys_kdma_name]
+            # If there is only one, make it a list for the following loop
+            if not isinstance(dset_files, list):
+                dset_files = [dset_files]
+            
+            # For each dataset file
+            for dset_f in dset_files:
+                with open(dset_f) as f:
+                    dset = json.load(f)
+                incontext_data[sys_kdma_name] = []
```

Comment:
If there are multiple dset files for one KDMA, I think this will overwrite with each new file. I **think** outside this for loop we'd want to do something like

```
if sys_kdma_name not in incontext_data:
    incontext_data[sys_kdma_name] = []
for dset_f in dset_files:
   ....
```
The if statement is even probably overkill

Apologies if that's not your intention with accepting a list of dset files per kdma now

Reply:
Good catch! Fixed here https://github.com/ITM-Kitware/align-system/pull/115/commits/48807cf3a519fba796786e9b845700dfae571dc6